compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -960,47 +989,327 @@ class JSCodeGen()(using genCtx: Context) {
-              jstpe.AnyType, mutable = false)
-          jsSuperClassParam :: params
+    /* organize constructors in a called-by tree
+     * (the implicit root is the primary constructor)
+     */
+    val ctorTree = {
+      val ctorToChildren = secondaryCtors
+        .groupBy(_.targetCtor)
+        .withDefaultValue(Nil)
+
+      /* when constructing the call-by tree, we use pre-order traversal to
+       * assign overload numbers.
+       * this puts all descendants of a ctor in a range of overloads numbers.
+       *
+       * this property is useful, later, when we need to make statements
+       * conditional based on the chosen overload.
+       */
+      var nextOverloadNum = 0
+      def subTree[T <: JSCtor](ctor: T): ConstructorTree[T] = {
+        val overloadNum = nextOverloadNum
+        nextOverloadNum += 1
+        val subtrees = ctorToChildren(ctor.sym).map(subTree(_))
+        new ConstructorTree(overloadNum, ctor, subtrees)
+      }
+
+      subTree(primaryCtor)
+    }
+
+    /* prepare overload dispatch for all constructors.
+     * as a side-product, we retrieve the capture parameters.
+     */
+    val (exports, jsClassCaptures) = {
+      val exports = List.newBuilder[jsExportsGen.Exported]
+      val jsClassCaptures = List.newBuilder[js.ParamDef]
+
+      def add(tree: ConstructorTree[_ <: JSCtor]): Unit = {
+        val (e, c) = genJSClassCtorDispatch(tree.ctor.sym, tree.ctor.params, tree.overloadNum)
+        exports += e
+        jsClassCaptures ++= c
+        tree.subCtors.foreach(add(_))
+      }
+
+      add(ctorTree)
+
+      (exports.result(), jsClassCaptures.result())
+    }
+
+    val (formalArgs, restParam, overloadDispatchBody) =
+      jsExportsGen.genOverloadDispatch(JSName.Literal("constructor"), exports, jstpe.IntType)
+
+    val overloadVar = js.VarDef(freshLocalIdent("overload"), NoOriginalName,
+        jstpe.IntType, mutable = false, overloadDispatchBody)
+
+    val ctorStats = genJSClassCtorStats(overloadVar.ref, ctorTree)
+
+    val constructorBody = js.Block(
+        paramVarDefs ::: List(overloadVar, ctorStats, js.Undefined()))
+
+    val constructorDef = js.JSMethodDef(
+        js.MemberFlags.empty,
+        js.StringLiteral("constructor"),
+        formalArgs, restParam, constructorBody)(OptimizerHints.empty, None)
+
+    (jsClassCaptures, constructorDef)
+  }
+
+  private def genPrimaryJSClassCtor(dd: DefDef): PrimaryJSCtor = {
+    val sym = dd.symbol
+    val Block(stats, _) = dd.rhs
+    assert(sym.isPrimaryConstructor, s"called with non-primary ctor: $sym")
+
+    var jsSuperCall: Option[js.JSSuperConstructorCall] = None
+    val jsStats = List.newBuilder[js.Tree]
+
+    /* Move all statements after the super constructor call since JS
+     * cannot access `this` before the super constructor call.
+     *
+     * scalac inserts statements before the super constructor call for early
+     * initializers and param accessor initializers (including val's and var's
```

Comment:
There's no early initializer in Scala 3.

Reply:
Fixed.